docs/specs/mlnet-database-loader/mlnet-database-loader-specs.md

```diff
@@ -0,0 +1,463 @@
+
+    - [ML.NET sample using millions of columns with sparse data](https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/getting-started/LargeDatasets)
+
+    SQL Server supports [Sparse columns](https://docs.microsoft.com/en-us/sql/relational-databases/tables/use-sparse-columns?view=sql-server-2017), however, it is just a way to optimize storage for null values. It still needs to have a real column created in the table per each logical column (i.e. 1,000 columns defined in the SQL table) even when it might not have data.
+
+
+
+
+
+
+# Gated conditions for first Preview release
+
+- TBD
+
+# Gated conditions for GA release
+
+- TBD
+
+# API design examples
+
+The following are a few C# examples on how this API could be used by a developer when loading data from a database into an IDataView.
+
+Whenever in the samples below it is using a `SqlConnection` type, it could also be using any other supported DbConnection in .NET providers, such as:
+
+- `SqlConnection` (SQL Server)
+- `NpgsqlConnection` (PostgreSQL)
+- `OracleConnection` (Oracle)
+- `MySqlConnection` (MySql)
+- etc.
+
+The specific type (`SqlConnection`, `NpgsqlConnection`, `OracleConnection`, etc.) is specified as a generic parameter so references to the needed NuGet packages will be provided by the user when trying to compile.
+
+**1. (Convenient 'must-have' method) Data loading from a database by specifying a SQL query sentence:**
+
+The signature of the `DataOperationsCatalog.LoadFromDbSqlQuery` method looks like:
+
+```C#
+public IDataView LoadFromDbSqlQuery<TRow, DbConnection>(string connString, string sqlQuerySentence) where T : class;
+```
+
+`TRow` is the model's input data class (Observation class) to be used by IDataView and ML.NET API.
+
+Example code using it:
+
+```C#
+MLContext mlContext = new MLContext();
+
+//Example loading from a SQL Server or SQL Azure database with a SQL query sentence
+IDataView trainingDataView = mlContext.Data.LoadFromDbSqlQuery<ModelInputData, SqlConnection>(connString: myConnString, sqlQuerySentence: "Select * from InputMLModelDataset where InputMLModelDataset.CompanyName = 'MSFT'"); 
+```
+
+**2. (Foundational method) Data loading from a database with an IDataReader object:**
+
+This is the foundational or pillar method which will be used by the rest of the higher level or convenient methods:
+
+The signature of the `DataOperationsCatalog.LoadFromDbDataReader` method looks like:
+
+```C#
+public IDataView LoadFromDbDataReader<TRow>(Func<IDataReader> executeReader) where TRow : class;
+```
+
+`TRow` is the model's input data class (Observation class) to be used by IDataView and ML.NET API.
+
+The following code is how the mlContext.Data.LoadFromDbDataReader() method is used:
+
+```C#
+MLContext mlContext = new MLContext();
+
+string conString = YOUR_CONNECTION_STRING;
+using (SqlConnection connection = new SqlConnection(conString))
```

Comment:
This is very interesting. 👍 Not sure if we want to go that path, but we'll evaluate it, for sure.
It might be interesting for our internal implementation code so we wouldn't need to take direct dependency on specific ADO.NET providers (as opposed to user's code where it is safer to take dependency to catch errors in compilation time). Thanks for these pointers to Orleans project! 👍

Reply:
@CESARDELATORRE No problem. To be clear, I don't mean you need to use that code or do something similar. Maybe more like give ideas how it could like for what I think is a fairly minimal way to create something that produces at the end strongly typed classes while supporting all the features the various databases can have.

I wrote that part of the code so I'm somewhat familiar with the problems. For instance, some drivers (or their versions) have bugs that they are not necessarily asynchronous. Or they just don't support streaming without some extra work. Maybe the largest thing here is that I indeed wouldn't like to see this tied to any specific ORM since what would happen is that I'd pipe plain SQL through it and use for serializing data and likely end up with some inefficiencies along the way (multiple resultsets?). I don't know if people want to bring their own ORMs or just write efficient SQL that likely involves analytic/windowing functions too.

Reply:
Hi Cesar,

I think that as a pre-requisite for this, we should fix IDataView LoadFromEnumerable so it works for all Enumerables. Developers can easily make Enumerables from almost anything - including any database.

Then we can create a set of helpers to create Enumerables from abstract 'Databases' (with concrete implementations). This sounds like the kind of problem that has been solved many times before? Once this is implemented then this can just feed the resulting Enumerables to LoadFromEnumerable... This is how AutoML would feed in training and test data (from databases or any kind of structured store?).

LoadFromEnumerable doesn't always work at the moment because code further on assumes that the Enumerable-backed IDataView is thread-safe - this is not always the case (ex. EF Core Context).  See: https://github.com/endintiers/Unearth.Demo.ML.FromDB 

I understand why limiting concurrency for trainers was a bad idea. We shouldn't. In the cases when training from huge amounts of data (in a DB for example) it is slooowww.

So I propose starting by adding a switch to LoadFromEnumerable 'singleThreaded=true' which would intercept IDataView cache update requests and force them onto a single thread. This may slow down the trainers a bit but nowhere near as much as single-threading them (where possible) and we can improve performance with eager loading.

Doing that first would remove a lot of complexity from the DatabaseLoader task (threading no longer an issue).

- Chris